### PR TITLE
Faster NB LL computation if data is known to be sparse

### PR DESCRIPTION
**Pull request by wukathy**
first try

### CODE REVIEW

File: scvi/distributions/_negative_binomial.py [hunk 1]
```diff
@@ -91,14 +93,27 @@ def log_nb_positive(x: torch.Tensor, mu: torch.Tensor, theta: torch.Tensor, eps=
 
     log_theta_mu_eps = torch.log(theta + mu + eps)
 
+    if is_sparse:
+        nonzero_x = x[x > 0]
```
Comment (adamgayoso) [inline scvi/distributions/_negative_binomial.py:97]: compute the mask once, store in a variable

File: scvi/distributions/_negative_binomial.py [hunk 2]
```diff
@@ -91,14 +93,27 @@ def log_nb_positive(x: torch.Tensor, mu: torch.Tensor, theta: torch.Tensor, eps=
...
+        terms_for_nonzero = (
+            theta * (torch.log(theta + eps) - log_theta_mu_eps)
```
Comment (adamgayoso) [inline scvi/distributions/_negative_binomial.py:99]:
```suggestion
```

Comment (adamgayoso) [reply scvi/distributions/_negative_binomial.py:99]: this is shared for zero and nonzero

File: scvi/distributions/_negative_binomial.py [hunk 3]
```diff
@@ -91,14 +93,27 @@ def log_nb_positive(x: torch.Tensor, mu: torch.Tensor, theta: torch.Tensor, eps=
...
+            + nonzero_x * (torch.log(mu + eps) - log_theta_mu_eps)
+            + torch.lgamma(nonzero_x + theta)
+            - torch.lgamma(theta)
+            - torch.lgamma(nonzero_x + 1)
```
Comment (adamgayoso) [inline scvi/distributions/_negative_binomial.py:103]: need to mask `mu`, `log_theta_mu_eps` with the nonzero mask.

the term `torch.lgamma(nonzero_x + theta)` might be trickier because theta has not been previously broadcast. Right here its shape is [1, n_genes], and the shape of nonzero_x will be a 1d tensor with all nonzero terms. So maybe theta should be initially broadcast (see Pyro code)

**Review by adamgayoso — CHANGES_REQUESTED** [review]
